openid/openid.go

```diff
@@ -0,0 +1,242 @@
+// Verify will use a cache that generates the discovery information in
+// the same way as Ubuntu SSO.
+var DefaultDiscoveryCache DiscoveryCache
+
+// Request contains the paramaters for an UbuntuSSO OpenID login request.
+type Request struct {
+	// ReturnTo contains the callback address for the RP, this is
+	// where the login response will come.
+	ReturnTo string
+
+	// Realm contains the realm that the user is logging in to. See
+	// http://openid.net/specs/openid-authentication-2_0.html#realms
+	// for details.
+	Realm string
+
+	// Teams contains a list of launchpad teams to query membership
+	// of for the logged in user.
+	Teams []string
+
+	// SimpleRegistrationRequired contains a list of simple
+	// registration fields that are required by the RP.
+	SimpleRegistrationRequired []string
+
+	// SimpleRegistrationOptional contains a list of simple
+	// registration fields that are optional, but requested by the
+	// RP.
+	SimpleRegistrationOptional []string
+}
+
+// RedirectURL creates an OpenID login request addressed to
+func RedirectURL(server usso.UbuntuSSOServer, r *Request) string {
+	v := url.Values{
+		"openid.ns":         {"http://specs.openid.net/auth/2.0"},
+		"openid.mode":       {"checkid_setup"},
+		"openid.claimed_id": {"http://specs.openid.net/auth/2.0/identifier_select"},
+		"openid.identity":   {"http://specs.openid.net/auth/2.0/identifier_select"},
+		"openid.return_to":  {r.ReturnTo},
+	}
+	if r.Realm != "" {
+		v.Set("openid.realm", r.Realm)
+	}
+	if len(r.Teams) > 0 {
+		v.Set("openid.ns.lp", nsTeams)
+		v.Set("openid.lp.query_membership", strings.Join(r.Teams, ","))
+	}
+	if len(r.SimpleRegistrationRequired) > 0 {
+		v.Set("openid.ns.sreg", nsSimpleRegistration)
+		v.Set("openid.sreg.required", strings.Join(r.SimpleRegistrationRequired, ","))
+	}
+	if len(r.SimpleRegistrationOptional) > 0 {
+		v.Set("openid.ns.sreg", nsSimpleRegistration)
+		v.Set("openid.sreg.optional", strings.Join(r.SimpleRegistrationOptional, ","))
+	}
+	return server.OpenIDURL() + "?" + v.Encode()
+}
+
+// Respones contains the values returned from Ubuntu SSO after a
+// successful login.
+type Response struct {
+	// ID contins the claimed_id of the logged in user. This will
+	// always be present in a successful login.
+	ID string
+
+	// Teams contains any launchpad teams, that were queried in the
+	// request, that the logged in user is a member of.
+	Teams []string
+
+	// SimpleRegistration contains any simple registration fileds are
+	// were assered in the response.
+	SimpleRegistration map[string]string
+}
+
+// verify is used to perform the OpenID verification of the login response.
+var verify = openid.Verify
+
+// Verify processes a positive assertion from Ubuntu SSO. If the
+// verification is successful any parameters asserted by Ubuntu SOO will
+// be set in the Response.
+func Verify(requestURL string, server usso.UbuntuSSOServer, ns NonceStore) (*Response, error) {
```

Comment:
Rather than having this as a bare function, I wonder if it might be best
as a method on a type that contains contextual information such as
the nonce store and the discovery cache (and even the UbuntuSSOServer?).

Then you probably wouldn't feel such a need for the global default variables.